gateway-service/src/main/java/org/zowe/apiml/gateway/security/config/ApimlX509AuthenticationFilter.java

```diff
@@ -35,26 +34,28 @@
 
     /**
      * Filter certificated by allowed - see publicKeyCertificatesBase64
+     *
      * @param certs all certificated to filter
      * @return filtered certificates (certs), contains only certificated used to sign in APIML
      */
-    private X509Certificate[] filter(X509Certificate[] certs) {
-        return Arrays.stream(certs)
-            .filter(x -> publicKeyCertificatesBase64.contains(
-                Base64.getEncoder().encodeToString(x.getPublicKey().getEncoded())
-            ))
-            .toArray(X509Certificate[]::new);
+    private Map<Boolean, List<X509Certificate>> filter(X509Certificate[] certs) {
+        return Arrays.stream(certs).collect(Collectors.partitioningBy(cer -> publicKeyCertificatesBase64.contains(Base64.getEncoder().encodeToString(cer.getPublicKey().getEncoded()))));
     }
 
     /**
      * Get certificates from request (if exists), filter them (to use only APIML certificate to request sign) and
      * store again into request.
+     *
      * @param request Request to filter certificates
      */
     private void filterCerts(ServletRequest request) {
         X509Certificate[] certs = (X509Certificate[]) request.getAttribute("javax.servlet.request.X509Certificate");
         if (certs != null) {
-            certs = filter(certs);
+            Map<Boolean, List<X509Certificate>> separatedCerts = filter(certs);
+            X509Certificate[] clientAuthCerts = new X509Certificate[separatedCerts.get(false).size()];
+            request.setAttribute("client.auth.X509Certificate", separatedCerts.get(false).toArray(clientAuthCerts));
+            X509Certificate[] zoweCerts = new X509Certificate[separatedCerts.get(true).size()];
+            certs = separatedCerts.get(true).toArray(zoweCerts);
```

Comment:
This class makes my brain hurt. The code is working but it could be made more readable, and the javadoc could use an update.

Reply:
maybe it's just me : )

Reply:
It's now refactored for readability